utils/optional_reference.hh

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <optional>
+#include <functional>
+
+namespace utils {
+
+template <class T>
+class optional_reference {
```

Comment:
kind of like https://en.cppreference.com/w/cpp/experimental/observer_ptr?

Reply:
Hmm, yeah, seems functionally equivalent.
I'll try to use that instead then.

Reply:
We already have a native language construct for this: a pointer. Why reinvent it?

Reply:
The main reason is the ability to initialize with a reference that's guaranteed not to be null, or with nullopt otherwise.
Also, since it's not a pointer, it can't be modified via pointer arithmetic and so on.

Reply:
You can do all this with a pointer. If you make your pointer `const T* const` you have all that you list above: a pointer that can be initialized from a reference or `nullptr` and cannot be modified via arithmetic and so on.

Reply:
It looks like `observer_ptr` provides that safe abstraction in a clear way.
Why not use it? Since it's experimental?

Reply:
@bhalevy I actually posted the link as "fun fact", it's probably safer to avoid experimental unless absolutely necessary. I actually like Botond's idea of a pointer with consts put in every possible corner.

Reply:
ok, if you guys insist...

Reply:
I hope you are trolling :grin:. Why do we want to suddenly replace good ol' language constructs with template classes?
I honestly very much dislike `std::reference_wrapper<T>` too, and think it only has place for covering a niche of binding references bind markers.

Reply:
> I honestly very much dislike std::reference_wrapper<T> too, and think it only has place for covering a niche of binding references bind markers.

_Just_ `reference_wrapper`? It's all over the `sharded<T>::start()` stuff! Is it this "binding references bind markers" case you mentioned?

Reply:
Yes, for example.

Reply:
I am not a fan of the home-grown optional_reference neither.